Accept input tensor with 0-dim batch size for MultiLabelMarginLoss

v0dro opened the pull request:
Fix for one of the layers listed in https://github.com/pytorch/pytorch/issues/12013 or https://github.com/pytorch/pytorch/issues/38115

v0dro committed:
working on supplying a good test

v0dro committed:
Merge branch 'master' of github.com:pytorch/pytorch into multi-label-margin-loss

v0dro committed:
Change MultiLabelMarginLoss to accept 0-dim batch sizes

v0dro committed:
move shape checking to separate function and reduce code duplication

v0dro committed:
accept 0 dim tensor in MultiLabelMargin

v0dro committed:
update NN modules

v0dro commented:
@ngimel request review.

v0dro commented:
@ngimel could you please have a look?

v0dro committed:
Merge branch 'master' of github.com:pytorch/pytorch into multi-label-margin-loss

v0dro committed:
Merge branch 'master' of github.com:pytorch/pytorch into multi-label-margin-loss

mruberry commented on test/test_nn.py:
```diff
@@ -10351,6 +10351,27 @@ def test_ReflectionPad_empty(self, device):
+        y = torch.ones(0, 10).type(torch.long)
+
+        out = mod(x, y)
+        out.sum().backward()
```
Check the value of out and x.grad, too.

mruberry commented:
Hey v0dro!

I made a comment about the test coverage. What about CUDA, by the way? Does it already handle inputs with a zero size batch dim?

v0dro committed:
Merge branch 'master' of github.com:pytorch/pytorch into multi-label-margin-loss

v0dro committed:
remove third party apps from previous commit

v0dro committed:
update multi label margin loss to work with JIT

v0dro committed:
remove include iostream

v0dro commented:
@mruberry could you please review this again?

ngimel commented on test/test_nn.py:
```diff
@@ -10351,6 +10351,32 @@ def test_ReflectionPad_empty(self, device):
             mod(inp)
 
+    @onlyOnCPUAndCUDA
+    def test_MultiLabelMarginLoss_empty(self, device):
```
you are fixing both MultiMarginLoss and MultiLabelMarginLoss, but you are testing only MultiLabelMarginLoss. 
Also, I did not see any changes to cuda, does cuda work already?

v0dro committed:
some extra tests

v0dro committed:
Merge branch 'master' of github.com:pytorch/pytorch into multi-label-margin-loss

v0dro committed:
update test for MultiLabel loss cehcks

v0dro committed:
change checks for GPU

v0dro committed:
update CUDA

v0dro committed:
update test for flake8

v0dro commented:
@mruberry @ngimel I've added tests for both modules (MultiLabelMarginLoss and MultiMarginLoss) and also updated the CUDA modules (I hadn't added a device option to the module constructor earlier).

Could you have a look again?

mruberry commented on aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu:
```diff
@@ -58,49 +58,52 @@ void THNN_(MultiLabelMarginCriterion_updateOutput)(
-          false
-          );
-      THCudaCheck(cudaGetLastError());
+    if (input->numel() != 0) {
```
It seems like this check should be `input->numel() == 0` and go around line 34 to short-circuit quickly and avoid the frees later. That would also minimize the diff size here.

v0dro replied:
Thanks for that. Have changed.

mruberry commented on aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu:
```diff
@@ -58,49 +58,52 @@ void THNN_(MultiLabelMarginCriterion_updateOutput)(
   }
   else {
-    TORCH_INTERNAL_ASSERT(false, "non-empty vector or matrix expected (shouldn't get here)");
+    TORCH_INTERNAL_ASSERT(false, "Expected 2D input with non-zero batch dim, or 1D input with non-zero dim, but got sizes: ", 
```
This doesn't seem correct because 2D can have zero batch dim (that's this PR) and 0D tensors also seem OK?

Maybe:

"Expected a tensor with two or fewer dimensions, but got X!"

Also TORCH_CHECK instead of TORCH_INTERNAL_ASSERT.

mruberry commented on aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu:
```diff
@@ -149,26 +152,29 @@ void THNN_(MultiLabelMarginCriterion_updateGradInput)(
-        gradInput->size(0), gradInput->size(1),
-        reduction == at::Reduction::Mean,
-        reduction != at::Reduction::None);
+    if (input->numel() != 0) {
```
Seems like this check should also live at the start of the function and short-circuit earlier.

v0dro replied:
Done.

mruberry commented on aten/src/THCUNN/generic/MultiMarginCriterion.cu:
```diff
@@ -59,78 +59,81 @@ void THNN_(MultiMarginCriterion_updateOutput)(
                "inconsistent target size");
-    dim3 blocks(input->size(0));
-    dim3 threads(MULTIMARGIN_THREADS);
+    if (input->numel() != 0) {
```
Again: can this check be inverted and a smaller short-circuit added for this case?

v0dro replied:
Done.

mruberry commented on aten/src/THCUNN/generic/MultiMarginCriterion.cu:
```diff
@@ -59,78 +59,81 @@ void THNN_(MultiMarginCriterion_updateOutput)(
   else
   {
-    AT_ERROR("non-empty vector or matrix expected, got sizes: ", input->sizes());
+    AT_ERROR("Expected 2D input with non-zero batch dim, or 1D input with non-zero dim, but got sizes: ", input->sizes());
```
Same updates needed here as mentioned above.

v0dro replied:
Done.

mruberry commented on aten/src/THCUNN/generic/MultiMarginCriterion.cu:
```diff
@@ -195,44 +198,46 @@ void THNN_(MultiMarginCriterion_updateGradInput)(
                "inconsistent target size");
-    dim3 blocks(gradInput->size(0));
-    dim3 threads(MULTIMARGIN_THREADS);
+    if (input->numel() != 0) {
```
Same question here about short-circuiting earlier.

v0dro replied:
Done.

mruberry commented on aten/src/THCUNN/generic/MultiMarginCriterion.cu:
```diff
@@ -195,44 +198,46 @@ void THNN_(MultiMarginCriterion_updateGradInput)(
   else
   {
-    AT_ERROR("non-empty vector or matrix expected, got ", input->sizes());
+    AT_ERROR("Expected 2D input with non-zero batch dim, or 1D input with non-zero dim, but got sizes: ", input->sizes());
```
Same updates here about error

v0dro committed:
Merge branch 'master' of github.com:pytorch/pytorch into multi-label-margin-loss

v0dro commented on test/test_nn.py:
```diff
@@ -10398,6 +10398,38 @@ def test_ReflectionPad_empty(self, device):
+
+            self.assertEqual(x, torch.zeros_like(x))
+
+            for p in mod.parameters():
```
Removed the loop.

v0dro commented on aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu:
```diff
@@ -149,26 +152,29 @@ void THNN_(MultiLabelMarginCriterion_updateGradInput)(
   }
   else {
-    AT_ERROR("non-empty vector or matrix expected, got size: ", gradInput->sizes());
+    AT_ERROR("Expected 2D input with non-zero batch dim, or 1D input with non-zero dim, but got sizes: ", gradInput->sizes());
```
Done.

v0dro commented on test/test_nn.py:
```diff
@@ -10398,6 +10398,38 @@ def test_ReflectionPad_empty(self, device):
+
+            with self.assertRaisesRegex(RuntimeError, 'Expected'):
+                x = torch.randn(0, requires_grad=True)
+                y = torch.ones(10).type(torch.long)
```
Done.

v0dro committed:
update testing and conditionals for testing

v0dro committed:
update tests

v0dro committed:
Merge branch 'master' of github.com:pytorch/pytorch into multi-label-margin-loss

v0dro committed:
chnage include headers to use angle brackets

v0dro committed:
revert use of angle brackets

v0dro committed:
update CUDA tests

v0dro committed:
multi margin loss shape check for CUDA

v0dro commented on test/test_nn.py:
```diff
@@ -10398,6 +10398,38 @@ def test_ReflectionPad_empty(self, device):
+                 torch.ones(0, 10, device=device).type(torch.long))]:
+
+            out = mod(x, y)
+            out.sum().backward()
```
Not using it throws an error when the tensor is empty.

This is the `out` object:
```
test/test_nn.py tensor([], device='cuda:0', dtype=torch.float32,
       grad_fn=<MultiMarginLossBackward
```

v0dro commented:
@ngimel could you please have a look at this PR again? CI failures are not related to my changes.

ngimel commented:
Can you please rebase to get a better test signal?

ngimel commented:
Reverted, it broke asan test.

ngimel commented on aten/src/ATen/native/LossMultiLabelMargin.cpp:
```diff
@@ -2,6 +2,7 @@
 #include <ATen/AccumulateType.h>
 #include <ATen/Dispatch.h>
 #include <ATen/TensorUtils.h>
+#include "LossMulti.h"
```
Please use standard include here, otherwise builds break (`<ATen/native/LossMulti.h>`)
